cpp/src/gandiva/regex_functions_holder.cc

```diff
@@ -275,4 +275,82 @@ const char* ExtractHolder::operator()(ExecutionContext* ctx, const char* user_in
   return result_buffer;
 }
 
+Result<std::shared_ptr<RegexpLikeHolder>> RegexpLikeHolder::Make(
+    const FunctionNode& node) {
+  ARROW_RETURN_IF(
+      node.children().size() < 2,
+      Status::Invalid("'regexp_like' function requires at least two parameters"));
+  ARROW_RETURN_IF(
+      node.children().size() > 3,
+      Status::Invalid("'regexp_like' function requires at most three parameters"));
+  auto pattern = arrow::internal::checked_cast<LiteralNode*>(node.children().at(1).get());
+  ARROW_RETURN_IF(
+      pattern == nullptr,
+      Status::Invalid(
+          "'regexp_like' function requires a literal as the second parameter"));
+
+  auto pattern_type = pattern->return_type()->id();
+  ARROW_RETURN_IF(
+      !(pattern_type == arrow::Type::STRING || pattern_type == arrow::Type::BINARY),
+      Status::Invalid(
+          "'regexp_like' function requires a string literal as the second parameter"));
+
+  if (node.children().size() > 2) {
+    auto parameter =
+        arrow::internal::checked_cast<LiteralNode*>(node.children().at(2).get());
+    if (parameter != nullptr) {
+      auto parameter_type = parameter->return_type()->id();
+      ARROW_RETURN_IF(
+          !arrow::is_binary_like(parameter_type),
+          Status::Invalid(
+              "'regexp_like' function requires a string literal as the third parameter"));
+      return RegexpLikeHolder::Make(std::get<std::string>(pattern->holder()), true,
+                                    std::get<std::string>(parameter->holder()));
+    }
+  }
+  return RegexpLikeHolder::Make(std::get<std::string>(pattern->holder()), false, "");
+}
+
+Result<std::shared_ptr<RegexpLikeHolder>> RegexpLikeHolder::Make(
+    const std::string& regex_pattern, bool used_match_parameter,
+    const std::string& match_parameter) {
+  RE2::Options regex_op;
+  // set re2 use posix regex expression which also called ERE in postgre sql
+  regex_op.set_posix_syntax(true);
+  // oracle's regex_like will default treat source str as single line
+  regex_op.set_one_line(true);
+
+  if (used_match_parameter) {
+    for (auto& parameter : match_parameter) {
+      switch (parameter) {
+        case 'i':
+          regex_op.set_case_sensitive(false);
+          break;
+        case 'c':
+          regex_op.set_case_sensitive(true);
+          break;
+        case 'n':
+          regex_op.set_dot_nl(true);
+          break;
+        case 'm':
+          regex_op.set_one_line(false);
+          break;
+        default:
+          ARROW_RETURN_NOT_OK(Status::Invalid("Invalid match parameter '", parameter,
+                                              "':Only 'i', 'c', 'n', 'm' are allowed"));
+      }
+    }
+  }
+
+  std::shared_ptr<RegexpLikeHolder> lholder;
+
+  lholder =
+      std::shared_ptr<RegexpLikeHolder>(new RegexpLikeHolder(regex_pattern, regex_op));
```

Comment:
* How about using `auto`?
* How about using just `holder`? Because we have only one holder in this context.

```suggestion
  auto holder =
      std::shared_ptr<RegexpLikeHolder>(new RegexpLikeHolder(regex_pattern, regex_op));
```